### ISSUE TITLE
Future context propagation behavior change between 1.X and 2.X

### ISSUE THREAD

**Issue by nadavwr** [kamon-io/Kamon#846]
We rely on Kamon 1.X for contextual logging. When upgrading we noticed some changes in behavior we rely on.

This [gist](https://gist.github.com/nadavwr/762a8cbf785b61110e92915c146dfcdd) contains two specs that are roughly equivalent (direct port of same code from 1.X to 2.X):
- Kamon1FutureSpec.scala
- Kamon2FutureSpec.scala

In Kamon 1.X, chaining Futures after leaving Kamon.withContextKey's scope discards  context.

Since Kamon 2.X, chaining Futures after leaving Kamon.runWithContextKey's scope retains the context.

Unfortunately for us, we rely on Kamon 1.X's behavior for adding logging context to Akka HTTP routes. The previous behavior enabled us to manage logging context for Akka HTTP's Route DSL by using the following directive:

```scala
val ContextKey: Context.Key[Map[String, String]] =
    Context.key("logging_context", Map.empty)

def currentParams(): Map[String, String] =
  Kamon.currentContext().get(LexicalContextKey)

def withLogCtx(params: (String, String)*): Directive0 =
  Directive { inner => ctx => 
    Kamon.runWithContextEntry(ContextKey, currentParams() ++ params) {
      inner(())(ctx)
    }
  }

// usage example:
  (path("/foo") & withLogCtx("root" -> "")) {
    concat(
      (withLogCtx("branch1" -> "") { reject },
      (withLogCtx("branch2" -> "") { complete(currentParams().keySet.mkString(",")) }
    )
  }
```

Suppose we send a `GET /foo` to the route above.
- We'd go info `/foo`, accumulating the `root` key into logging context
- We'd go into the first route in the concat, thereby accumulating `branch1`
- We'd get rejected by the first route in the concat. Here's where the versions differ:
  - Kamon 1.X would have popped "branch1" from the context—kinda like lexical scope
  - Kamon 2.X retains "branch1" in context
- After having been rejected we proceed the next inner route, accumulating "branch2"
- We then return the set of all keys in context

In Kamon 1.X we used to see `root,branch2`. Kamon 2.X gives us `root,branch1,branch2`, even though we obviously weren't expecting `branch1` in the context.

Versions for Kamon 1.X:
- kamon-core 1.1.6
- kamon-executors 1.0.1
- kamon-scala-future 1.0.0
- aspectj weaver 1.9.2

Versions for Kamon 2.X:
- kamon-core 2.13
- kamon-bundle 2.13
- Unit tests are run with -javaagent, actual application invokes `Kamon.init()` via early initializer in order to precede classloading

We'd appreciate guidance on how we can regain the original functionality, or approximate it on Kamon 2.X. Even shedding light on any discussion occurring around the time of this change might help give us a lead.

Thanks!

**Comment by danielkman** [comment]
👍

**Comment by nadavwr** [comment]
Reproducer is now available [here](https://github.com/nadavwr/kamon-future-repro).
The test for Kamon 1.X succeeds, while the test for Kamon 2.X fails.
I've added CircleCI integration so you can see it in action.

**Comment by ivantopo** [comment]
Hey @nadavwr, thanks for the reproducible and the very detailed explanation! 

I used the reproducible [you shared](https://github.com/nadavwr/kamon-future-repro) to try figuring out what can be done here. First thing is, do you maybe have a reproducible that uses the Akka HTTP directives you mentioned above? I understand how the example you shared works when it is Future-based, but I really don't get how the issue manifests in Akka HTTP with the code you shared above because there are no Futures anywhere there, or at least I don't see it just from the example code :thinking: 

We have a few open issues related to Futures and I'm not sure what is going to be the proper solution for all of them, but I can offer you a workaround to get the exact same behavior from Kamon 1.x on Kamon 2.x with this configuration settings:

```
kanela.modules {
  scala-future {
    enabled = no
  }

  executor-service {
    within += "scala.concurrent.impl.CallbackRunnable"
    exclude = [
      "^java.*",
      "^sun.*",
      "^com.sun.tools.*",
      "^sbt.internal.*",
      "^com.intellij.rt.*",
      "^org.jboss.netty.*",
      "^com.google.common.base.internal.Finalizer",
      "^kamon.module.*",
      "^kamon.instrumentation.executor.ExecutorInstrumentation.*",
      "^kanela.agent.*",
      "com.twitter.util.ConstFuture.*"
    ]
  }
}
```

I tried your reproducible with the latest Kamon version (v2.1.20 at the moment) with those settings and it worked as expected. Just keep in mind that you should not use the utility functions from `kamon.instrumentation.futures.scala.ScalaFutureInstrumentation` because they will leave dirty threads for sure.

Stay tuned around the Future-related issues as @SimunKaracic and I are working to get a proper solution to all of this. Thanks for your patience!

**Comment by ivantopo** [comment]
Hey folks, this is fixed and released with Kamon [2.2.0](https://github.com/kamon-io/Kamon/releases/tag/v2.2.0). Please share if you find any issues with the new version!
